drivers/spi/spi_ambiq_spid.c

```diff
@@ -0,0 +1,432 @@
+			data->ios_cfg.ui32InterfaceSelect |= AM_HAL_IOS_SPIMODE_1;
+		} else {
+			data->ios_cfg.ui32InterfaceSelect |= AM_HAL_IOS_SPIMODE_0;
+		}
+	}
+
+	if (config->operation & SPI_OP_MODE_MASTER) {
+		LOG_ERR("Controller mode not supported");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_MODE_LOOP) {
+		LOG_ERR("Loopback mode not supported");
+		return -ENOTSUP;
+	}
+
+	if (spi_cs_is_gpio(config)) {
+		LOG_ERR("CS control via GPIO is not supported");
+		return -EINVAL;
+	}
+
+	/* Eliminate the "read-only" section, so an external controller can use the
+	 * entire "direct write" section.
+	 */
+	data->ios_cfg.ui32ROBase = AMBIQ_SPID_FIFO_BASE;
+	/* Making the "FIFO" section as big as possible. */
+	data->ios_cfg.ui32FIFOBase = AMBIQ_SPID_FIFO_BASE;
+	/* We don't need any RAM space, so extend the FIFO all the way to the end
+	 * of the LRAM.
+	 */
+	data->ios_cfg.ui32RAMBase = AMBIQ_SPID_FIFO_END;
+	/* FIFO Threshold - set to half the size */
+	data->ios_cfg.ui32FIFOThreshold = AMBIQ_SPID_FIFO_LENGTH >> 1;
+
+	data->ios_cfg.pui8SRAMBuffer = ambiq_spid_sram_buffer,
+	data->ios_cfg.ui32SRAMBufferCap = AMBIQ_SPID_TX_BUFSIZE_MAX,
+
+	ctx->config = config;
+
+	ret = am_hal_ios_configure(data->ios_handler, &data->ios_cfg);
+
+	return ret;
+}
+
+static int spi_ambiq_xfer(const struct device *dev, const struct spi_config *config)
+{
+	struct spi_ambiq_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	int ret = 0;
+	uint32_t chunk, num_written, num_read, used_space = 0;
+
+	while (1) {
+		/* First send out all data */
+		if (spi_context_tx_on(ctx)) {
+			spi_ambiq_inform(dev);
+			chunk = (ctx->tx_len > AMBIQ_SPID_TX_BUFSIZE_MAX)
+					? AMBIQ_SPID_TX_BUFSIZE_MAX
+					: ctx->tx_len;
+			am_hal_ios_fifo_space_used(data->ios_handler, &used_space);
+			/* Controller done reading the last block signalled
+			 * Check if any more data available
+			 */
+			if (!used_space) {
+				if (ctx->tx_buf) {
+					/*  Copy data into FIFO */
+					ret = am_hal_ios_fifo_write(data->ios_handler,
+								    (uint8_t *)ctx->tx_buf, chunk,
+								    &num_written);
+				} else {
+					uint32_t dummy_written = 0;
+
+					num_written = 0;
+					/*  Copy dummy into FIFO */
+					while (chunk) {
+						uint32_t size = (chunk > AMBIQ_SPID_DUMMY_LENGTH)
+									? AMBIQ_SPID_DUMMY_LENGTH
+									: chunk;
+						ret = am_hal_ios_fifo_write(
+							data->ios_handler,
+							ambiq_spid_dummy_buffer[0], size,
+							&dummy_written);
+						num_written += dummy_written;
+						chunk -= dummy_written;
+					}
+				}
+				if (ret != 0) {
+					LOG_ERR("SPID write error: %d", ret);
+					goto end;
+				}
+				spi_context_update_tx(ctx, 1, num_written);
+			}
+		} else if (spi_context_rx_on(ctx)) {
+			/* Wait for controller write complete interrupt */
+			(void)k_sem_take(&data->spim_wrcmp_sem, K_FOREVER);
+			/* Read out the first byte as packet length */
+			num_read = am_hal_ios_pui8LRAM[0];
+			if (ctx->rx_buf) {
+				if (ctx->rx_len < num_read) {
```

Comment:
You need to look how other drivers works, and how spi buffers are manipulated.
This is not right in 2 aspects:
 - because the current spi_buf may be too small for the whole, but the spi_buf_set may have another spi_buf to fill in after this current one. 
 - from a driver perspective you don't have to interpret if the user made a mistake providing enough space or not. It's a user problem, not a driver one. It could be that the user is only interested by some bytes and does not care about the rest.

Reply:
Removed rx_buf size check and added polling for multiple spi_buf, please help to review again, thank you!